lib/resources/users.js

```diff
@@ -78,29 +79,23 @@ module.exports = (service, { Actee, User, mail }) => {
       .then((user) => auth.canOrReject('read', user.actor)
         .then(() => user))));
 
-  /* The following endpoints are not part of v1 scope. so rather than expose hidden
-   * surfaces that will not be well-tested, we shall simply comment them out for now.
-
   // TODO: infosec debate around 404 vs 403 if insufficient privs but record DNE.
-  service.put('/users/:id', endpoint(({ params, body, auth }) =>
+  service.patch('/users/:id', endpoint(({ params, body, auth }) =>
     User.transacting().getByActorId(params.id)
       .then(getOrNotFound)
       .then((user) => auth.canOrReject('update', user.actor)
-        .then(() => {
-          const merged = user.with(User.fromApi(body));
-          return merged.update().then(() => merged);
-        }))));
+        .then(() => user.with(User.fromApi(without([ 'id' ], body))).update())
```

Comment:
so if you don't strip the `id` _before_ `fromApi`, then the `id` given by the user ends up on `actor.id`, so then the database tries to update that record instead of the one it's supposed to.

what's actually happening (i think) is not that `user.with` overwrites `user.actorId` with `undefined`, but rather than because there is no `id`, the extant `user.actorId` is preserved, which is what we want.

keeping those fields in sync is not an option. the more likely correct answer is to have `User.fromApi` never accept `id` in the first place, since there should be zero cases in which an API consumer should successfully communicate that value.

Reply:
> the more likely correct answer is to have `User.fromApi` never accept `id` in the first place, since there should be zero cases in which an API consumer should successfully communicate that value.

An approach like that sounds good to me!

I added some logging to the endpoint, and I'm seeing the following values when I update my `displayName` from Matt White to Matt Whit:

```
console.log(user);

 instance {
  actorId: 5,
  password: 'redacted',
  mfaSecret: null,
  email: 'a@s.t',
  actor:
   instance {
     id: 5,
     type: 'user',
     acteeId: '6a5887ea-7dc0-4cfb-8d84-0f08d8972b30',
     displayName: 'Matt White',
     meta: null,
     createdAt: 2018-06-26T15:46:19.942Z,
     updatedAt: 2018-10-19T19:43:06.532Z,
     deletedAt: null,
     systemId: null } }

console.log(User.fromApi(without([ 'id' ], body)));

 instance {
  email: 'a@s.t',
  actor: instance { displayName: 'Matt Whit', type: 'user' },
  actorId: undefined }

console.log(user.with(User.fromApi(without([ 'id' ], body))));

 instance {
  actorId: undefined,
  password: 'redacted',
  mfaSecret: null,
  email: 'a@s.t',
  actor:
   instance {
     id: 5,
     type: 'user',
     acteeId: '6a5887ea-7dc0-4cfb-8d84-0f08d8972b30',
     displayName: 'Matt Whit',
     meta: null,
     createdAt: 2018-06-26T15:46:19.942Z,
     updatedAt: 2018-10-19T19:43:06.532Z,
     deletedAt: null,
     systemId: null } }
```

From what I can tell, it seems like `user.actorId` is being overwritten with `undefined`. On the other hand, `user.actor.id` is unchanged, and I think that's the property that the endpoint actually uses. I think the update you suggest to `User.fromApi()` should resolve this case, because with that change, the object that `User.fromApi()` returns wouldn't have an `actorId` property, so it wouldn't overwrite any existing `actorId` property.

Alternatively, if `User.fromApi()` continues to accept `id`, maybe `User.fromApi()` could be updated so that the object it returns has an `actorId` property only if the `data` argument has an `id` property? I'm thinking of something along these lines:

```js
static fromApi(data) {
  const userData = pick(User.fields(), data);
  userData.actor = new Actor(merge(pick(Actor.fields(), data), { type: 'user' }));
  if ('id' in data) userData.actorId = data.id;
  return new User(userData);
}
```

Reply:
hm, something must have drifted since i did a similar inquiry. either way, i think i'm revisiting this at a broader level soon here.

Reply:
as mentioned, i made `fromApi` not accept `id` ever